File: code/week-01/java/PerEncoding.java
/* Week 1 -- Introduction to Data Structures
 * A miniature PER-style encoding: no tags, no length for the fixed-size
 * name field, and age packed into just the bits its range needs.
 * CEN207 Data Structures (CS50-style lecture notes)
 */
public class PerEncoding {
    static int bitpos;

    /* Pack the low `width` bits of `value` into buf, starting at bit offset bitpos (MSB first). */
    static void packBits(byte[] buf, int value, int width) {
        for (int i = width - 1; i >= 0; i--) {
            int bit = (value >> i) & 1;
            int byteIndex = bitpos / 8;
            int bitIndex = 7 - (bitpos % 8);
            if (bit != 0)
                buf[byteIndex] |= (byte) (1 << bitIndex);
            bitpos++;
        }
    }

    public static void main(String[] args) {
        byte[] per = new byte[4];   // 32 bits: 24 (name) + 5 (age) + 3 padding
        bitpos = 0;
        String name = "Rex";

        for (int i = 0; i < 3; i++)
            packBits(per, name.charAt(i), 8);   // fixed size: no length needed
        packBits(per, 5, 5);                     // age, constrained to 0..31: 5 bits

        StringBuilder sb = new StringBuilder();
        for (byte b : per)
            sb.append(String.format("%02X ", b));
        System.out.println("PER: " + bitpos + " significant bits (no tags, no length for name, age in 5 bits), packed into "
                + ((bitpos + 7) / 8) + " bytes: " + sb.toString().trim());
        System.out.println("BER (Section 6.5) used 10 bytes (80 bits) for the same record.");
    }
}

File: code/week-01/c/per_encoding.c
/* Week 1 -- Introduction to Data Structures
 * A miniature PER-style encoding: no tags, no length for the fixed-size
 * name field, and age packed into just the bits its range needs.
 * CEN207 Data Structures (CS50-style lecture notes)
 */
#include <stdio.h>

/* Pack the low `width` bits of `value` into buf, starting at bit offset *bitpos (MSB first). */
static void pack_bits(unsigned char *buf, int *bitpos, unsigned int value, int width) {
    for (int i = width - 1; i >= 0; i--) {
        int bit = (int) ((value >> i) & 1u);
        int byte_index = *bitpos / 8;
        int bit_index = 7 - (*bitpos % 8);
        if (bit)
            buf[byte_index] |= (unsigned char) (1u << bit_index);
        (*bitpos)++;
    }
}

int main(void) {
    unsigned char per[4] = {0};   /* 32 bits: 24 (name) + 5 (age) + 3 padding */
    int bitpos = 0;
    const char name[] = "Rex";

    for (int i = 0; i < 3; i++)
        pack_bits(per, &bitpos, (unsigned char) name[i], 8);   /* fixed size: no length needed */
    pack_bits(per, &bitpos, 5, 5);                              /* age, constrained to 0..31: 5 bits */

    printf("PER: %d significant bits (no tags, no length for name, age in 5 bits), packed into %d bytes:",
           bitpos, (bitpos + 7) / 8);
    for (int i = 0; i < 4; i++)
        printf(" %02X", per[i]);
    printf("\n");

    printf("BER (Section 6.5) used 10 bytes (80 bits) for the same record.\n");

    return 0;
}
